### PR TITLE
Documentation generation

### PR DESCRIPTION
**Pull request by zainsohail04**
#32 suggests to implement an automated way for documentation. So far, it is automated to the level of each module within the package. We need to manually add an rst file for each module that is added. I have automated the arg types so that typehints are used to automatically include that information, as well as the default values. One thing to keep in mind is that the return docstring doesn't support multi value output so it has to be wrapped within a tuple.

The matplotlib dependency is also within the pull request.

### CODE REVIEW

File: sed/binning.py [hunk 1]
```diff
@@ -25,36 +24,36 @@ def _arraysum(array_a, array_b):
...
+    nBins: int | Sequence[int] = 100,
     hist_mode: str = "numba",
     jitterParams: dict = None,
     return_edges: bool = False,
-) -> np.ndarray:
+) -> np.ndarray | tuple[np.ndarray, list]:
     """Compute the n-dimensional histogram of a single dataframe partition.
 
     Args:
-        part (ddf.DataFrame): dataframe on which to perform the histogram.
```
Comment (daviddoji) [inline sed/binning.py:40]: why types were removed everywhere?

File: sed/binning.py [hunk 2]
```diff
@@ -1,7 +1,6 @@
 # All functions in this file are adapted from https://github.com/mpes-kit/mpes
 from functools import reduce
 from typing import Sequence
-from typing import Tuple
```
Comment (daviddoji) [inline sed/binning.py:4]: if we stick to Python 3.8, this import is needed

File: pyproject.toml [hunk 3]
```diff
@@ -17,6 +17,11 @@ xarray = "^0.21.1"
 numba = "^0.55.1"
 threadpoolctl = "^3.1.0"
 tqdm = "^4.62.3"
```
Comment (daviddoji) [inline pyproject.toml:19]: That's ok, but next time make another PR if two different issues are addressed, e.g. addind new dependency and creating documentation.
It's easier for review ;)

File: sed/binning.py (hunk 1 above)
Comment (zainsohail04) [reply sed/binning.py:40]: As mentioned above 

> I have automated the arg types so that typehints are used to automatically include that information, as well as the default values.

There is no longer a need for including the types twice henceforth.

File: pyproject.toml (hunk 3 above)
Comment (zainsohail04) [reply pyproject.toml:19]: Right. Most of the dependencies added are for documentation purposes. I saw that you included matplotlib in the other PR so I think it can be removed from here.

File: sed/binning.py (hunk 2 above)
Comment (daviddoji) [reply sed/binning.py:4]: Although `pyenv` makes it really easy to choose any Python version, we should stick not to the bleeding-edge ones. 
I chose 3.8 because it is the one installed on Maxwell.

Comment (rettigl) [reply sed/binning.py]: Do you use 3.8 features that are not part of 3.7? I would stick to the lowest version that provides the features necessary for the code. The old packages run just fine in 3.7. I would not break this compatibility without very good reason

Comment (zainsohail04) [reply sed/binning.py]: > Do you use 3.8 features that are not part of 3.7? I would stick to the lowest version that provides the features necessary for the code. The old packages run just fine in 3.7. I would not break this compatibility without very good reason

Currently not in use but the importlib.metadata can end up being quite useful in many situations. For example to use in single source versioning throughout the package. 
I am not aware if any other features from 3.8 are or will be used. With 3.10, I already had some reason to have used it.

Comment (daviddoji) [reply sed/binning.py]: > Do you use 3.8 features that are not part of 3.7? I would stick to the lowest version that provides the features necessary for the code. The old packages run just fine in 3.7. I would not break this compatibility without very good reason

I chose 3.8 for two reasons:
- It's the default in Maxwell when you spawn a node
- Some dependencies (pandas, dask, ...) require, at least, python 3.8 in their last releases. As in some point, we will need to update it anyway, I decided to do it the sooner the better considering this library is still under heavy development

Comment (steinnymir) [reply sed/binning.py]: I also agree that python 3.8 is the better choice. 
Earlier versions will only cause us to need to update sooner than later.
Bleeding edge versions like 3.10 are not easy to handle on the user side, outweighting the advantages posed by the new features.

File: sed/binning.py (hunk 1 above)
Comment (steinnymir) [reply sed/binning.py:40]: This will be a little unconfortable, but I guess its ok.

I usually autogenerate the docstring, and this includes the types by default. Maybe I can find some setting that changes that.

Comment (steinnymir) [reply sed/binning.py:40]: I thought this would be unconfortable when using autogenerated docstring strucutres, but it turns out there is a setting to use google-notypes (or whatever base style) which does the job!

**Review by steinnymir — APPROVED** [review]
All looks good

**Comment by zainsohail04** [comment]
> All looks good to me, so once the conflict is resolved, feel free to merge from my side. @rettigl and @zainsohail04? your reviews are still open, do you agree with the changes proposed?

Since I am the author, I can not also be the reviewer of this. However, I agreed with @daviddoji and changed the version to 3.8 already.

**Review by daviddoji — APPROVED** [review]

**Comment by daviddoji** [comment]
> > All looks good to me, so once the conflict is resolved, feel free to merge from my side. @rettigl and @zainsohail04? your reviews are still open, do you agree with the changes proposed?
> 
> Since I am the author, I can not also be the reviewer of this. However, I agreed with @daviddoji and changed the version to 3.8 already.

I think you'll need to checkout the `poetry.lock` and `pyproject.toml` files to resolve the conflicts

**Review by rettigl — APPROVED** [review]
I don't overlook all changes here, but sure, go ahead. One further comment: Does in addition to the poetry stuff the standard pip installation still work?

**Comment by daviddoji** [comment]
> I don't overlook all changes here, but sure, go ahead. One further comment: Does in addition to the poetry stuff the standard pip installation still work?

Do you mean installing the dependencies from a `requirements.txt` files? That can easily be done, if people find it useful.
Do you mind @rettigl to open an issue so we keep track of this request?
